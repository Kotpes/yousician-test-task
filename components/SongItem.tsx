import React from 'react'
import {Song} from '../hooks/useFetch'
import HeartIcon from './HeartIcon'

import styles from './SongItem.module.css'

interface Props extends Song {
    index: number
    isLiked: boolean
}

const SongItem = (props: Props) => {
    const {images, title, artist, level, isLiked = false, index} = props
    const containerBackgroundStyle = index % 2 == 0 ? styles.even : ''

    //adds fallback image for those images that return 403
    const handleFallbackImage = (event: any) => {
        event.target.src = '/missing-cover.png'
    }

    return (
        <section className={`${styles.container} ${containerBackgroundStyle}`}>
            <img src={images} onError={handleFallbackImage} alt={title} className={styles.coverArt}/>
            <div className={styles.songInfo}>
                <h3 className={styles.title}>{title}</h3>
                <h4 className={styles.artist}>{artist}</h4>
            </div>
            <div className={styles.songLevelAndAction}>
                {level}
                <HeartIcon fill={isLiked ? '#dc001d' : 'none'} stroke={isLiked ? 'none' : '#fff'} />
            </div>
        </section>
    )
}

export default SongItem