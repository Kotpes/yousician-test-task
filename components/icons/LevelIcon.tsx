import inRange from 'lodash.inrange'

const getColor = (level: number) => {
    //1-5 green, 6-10 orange, 10-15 red
    if (inRange(level, 1, 6)) {
        return '#6fc13e'
    }
    if (inRange(level, 6, 11)) {
        return '#ff8e00'
    }
    if (inRange(level, 10, 16)) {
        return '#dc001c'
    }
}

interface Props {
    value: number
    isSelected?: boolean
}

const SvgComponent = ({ value = 1, isSelected }: Props) => {
    const radius = 15;
    const segmentWidth = value * 100 / 15
    const circumference = Math.PI * (radius * 2);
    const progress = ((100 - segmentWidth) / 100) * circumference;
    const progressColor = getColor(value)
    const fill = isSelected ? '#fff' : 'transparent'
    const textColor = isSelected ? '#000' : '#fff'
    const progressStrokeColor = isSelected ? '#fff' : progressColor
    const deviderStrokeColor = isSelected ? '#fff' : '#000'
    const ringStrokeColor = isSelected ? '#fff' : '#383635'

    return (
        <svg viewBox="0 0 42 42" width={45} height={45}>
            <circle
                className="ring"
                cx={21}
                cy={21}
                r={radius}
                fill={fill}
                stroke={ringStrokeColor}
                strokeDasharray={0}
                strokeWidth={1.8}

            />
            <circle
                className="progress"
                cx={21}
                cy={21}
                r={radius}
                fill={fill}
                stroke={progressStrokeColor}
                strokeDasharray={circumference}
                strokeDashoffset={progress}
                strokeWidth={1.8}
                transform='rotate(-90 21 21)'
            />
            <circle
                className="devider"
                cx={21}
                cy={21}
                r={radius}
                fill={fill}
                stroke={deviderStrokeColor}
                strokeDasharray="9% 65.5%"
                transform='rotate(-96 21 21)'
                strokeWidth={2}
            />
            <text
                x="50%"
                y="50%"
                textAnchor="middle"
                fill={textColor}
                dy=".3em"
                fontSize="14"
                fontWeight="800"
            >
                {value}
            </text>
        </svg>
    )
}

export default SvgComponent