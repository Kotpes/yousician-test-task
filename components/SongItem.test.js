import { create, act } from 'react-test-renderer';

import SongItem from './SongItem';
import HeartIcon from './icons/HeartIcon';
import LevelIcon from '../components/icons/LevelIcon';

const props = {
  images: 'images_url',
  title: 'Song title',
  artist: 'Artist name',
  level: 10,
  isLiked: false,
  id: 'songId',
  reFetchFavourites: jest.fn(() => null),
  favorite: { id: 'favouriteId', songId: 'songId' },
  index: 1,
};

fetch.mockResponseOnce(JSON.stringify({}));

describe('SongItem', () => {
  test('matches the snapshot', () => {
    const component = create(<SongItem {...props} />);

    expect(component.toJSON()).toMatchSnapshot();
  });
  test('adds correct props to HeartIcon after adding song to favorites', () => {
    let component;
    act(() => {
      component = create(<SongItem {...props} />);
    });
    const instance = component.root;
    const heartIconButton = instance.findByProps({ role: 'button' });
    const heartIcon = instance.findByType(HeartIcon);
    expect(heartIcon.props).toEqual({ fill: 'none', stroke: '#fff' });

    act(() => {
      heartIconButton.props.onClick();
    });

    expect(heartIcon.props).toEqual({ fill: '#dc001d', stroke: 'none' });
  });
  test('LevelIcon has correct props', () => {
    const component = create(<SongItem {...props} />);
    const instance = component.root;
    const levelIconComponent = instance.findByType(LevelIcon);

    expect(levelIconComponent.props.value).toEqual(10);
  });
});
