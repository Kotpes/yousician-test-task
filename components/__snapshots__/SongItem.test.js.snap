// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SongItem matches the snapshot 1`] = `
<section
  className="container "
>
  <img
    alt="Song title"
    className="coverArt"
    onError={[Function]}
    src="images_url"
  />
  <div
    className="songInfo"
  >
    <h3
      className="title"
    >
      Song title
    </h3>
    <h4
      className="artist"
    >
      Artist name
    </h4>
  </div>
  <div
    className="songLevelAndAction"
  >
    <span
      className="songLevel"
    >
      <svg
        height={45}
        viewBox="0 0 42 42"
        width={45}
      >
        <circle
          className="ring"
          cx={21}
          cy={21}
          fill="transparent"
          r={15}
          stroke="#383635"
          strokeDasharray={0}
          strokeWidth={1.8}
        />
        <circle
          className="progress"
          cx={21}
          cy={21}
          fill="transparent"
          r={15}
          stroke="#ff8e00"
          strokeDasharray={94.24777960769379}
          strokeDashoffset={31.41592653589792}
          strokeWidth={1.8}
          transform="rotate(-90 21 21)"
        />
        <circle
          className="devider"
          cx={21}
          cy={21}
          fill="transparent"
          r={15}
          stroke="#000"
          strokeDasharray="9% 65.5%"
          strokeWidth={2}
          transform="rotate(-96 21 21)"
        />
        <text
          dy=".3em"
          fill="#fff"
          fontSize="14"
          fontWeight="800"
          textAnchor="middle"
          x="50%"
          y="50%"
        >
          10
        </text>
      </svg>
    </span>
    <span
      onClick={[Function]}
      onKeyPress={[Function]}
      role="button"
      tabIndex={0}
    >
      <svg
        height={24}
        style={
          Object {
            "cursor": "default",
          }
        }
        viewBox="0 0 24 24"
        width={24}
      >
        <path
          d="M0 0h24v24H0z"
          fill="none"
        />
        <path
          d="M12 21.35l-1.45-1.32C5.4 15.36 2 12.28 2 8.5 2 5.42 4.42 3 7.5 3c1.74 0 3.41.81 4.5 2.09C13.09 3.81 14.76 3 16.5 3 19.58 3 22 5.42 22 8.5c0 3.78-3.4 6.86-8.55 11.54L12 21.35z"
          fill="none"
          stroke="#fff"
        />
      </svg>
    </span>
  </div>
</section>
`;
